enhancements/ignition-spec-dual-support.md

```diff
@@ -0,0 +1,327 @@
+this case, as well as how to mitigate.
+
+** How to properly detect which ignition version to serve new machines **
+
+This will likely be somewhat difficult to handle, as the wrong version on a 
+system will completely break the boot process for a new machine. To mitigate, 
+the method of detection should either be made very robust, or we should 
+consider not allowing users to use ignition v0s2 bootimages after the update 
+to spec 3. This should also be considered in conjunction with functionality 
+to update bootimages, which will make this easier.
+
+
+## Design Details
+
+The implemented changes for the various components can be separate, with the
+caveat that ignition spec 3 support for MCO must happen first (so that other
+component changes can be tested in cluster). The MCO changes can be standalone,
+as they serve to bring a spec 2 cluster to spec 3, or work as in on a spec 3
+cluster.
+
+** RHCOS details: **
+
+RHCOS must change to use ignition v2, which supports spec 3 configs. The actual
+switching of the package is very easy on RHCOS. The building of ignition v2,
+however, presents two issues:
+
+ - The util-linux package is old on RHEL, without support of “lsblk -o PTUUID” which ignition uses. This will have to be reworked in RHCOS, or the package must be bumped and rebuilt for rhel 8.1 or workaround as in https://github.com/coreos/ignition-dracut/pull/133
+ - Ignition-dracut has seen significant deltas between the spec2x and master (spec 3) branches, especially for initramfs network management. There are also minor details such as targets that need to be checked for existing units. There exists a need to merge some spec2x bits back into 3x before RHCOS can move to 3x
+
+** Installer details: **
+
+The installer would only need to support EITHER ignition spec 2 or spec 3, but
+does not have to support both simultaneously. At the time of writing this
+proposal, there exist FCOS/OKD branches for the installer that are looking to
+move to spec 3, and has had success in installing a cluster. This work can be
+integrated for OCP as well. The main issue remaining is due to the necessity of
+moving to go modules as the vendoring method, there are failures in the Azure
+Terraform provider that seem to be incompatible with this change.
+
+** MCO details: **
+
+A spec translator will first be implemented in the MCO, with the ability to
+detect untranslatable configs. The MCO then should be updated to have support
+for both ignition V0S2 and V2S3. Since the MCO is responsible for the current
+cluster nodes, it will be the only place at which spec translation is done.
+The translation will happen when the version of MCO with dual support and
+translator is first deployed; it will detect the existing config being spec 2,
+generate a new renderedconfig based on a translator from spec2 to spec 3, and
+do a “dummy reboot” (if necessary) to bring the cluster to the desired config,
```

Comment:
What exactly do you mean by "dummy reboot" in this scenario? When would the "dummy reboot" be necessary?

Reply:
Since the cluster machineconfig is saved as a "desiredConfig" hash, I'm not sure the best way to translate that yet. This just means that after the specs are translated, if we need, we can set a new desiredConfig that specifies a successful translation to spec 3, which the cluster needs to reboot to bring the currentConfig up to.